### PR TITLE
InSpec Waivers: --filter-waived-controls

### PR DESCRIPTION
**Pull request by Schwad**
This PR is being adopted and brought up to a releasable state. Along the way we're probably going to drop the "ludicrous" term for something more formal :-), like `--filter-waived-controls`.

Major things include adding error checking, making the regexes less sensitive to accidental errors, and some refactoring/renaming.
@clintoncwolfe

----

This is an experimental PR to investigate alternative, optional flags for ensuring that waived controls are not evaluated at all. This could be considered __Ludicrous Speed__ for scaled customers.

ludicrous

## Background

**Links for background**

* https://github.com/inspec/inspec/issues/5068
* https://github.com/inspec/inspec/issues/3760
* https://github.com/inspec/inspec/issues/5225
* https://github.com/inspec/inspec/issues/5215
* https://github.com/inspec/inspec/issues/5215

TL;DR: Right now all control code outside describe blocks is silently evaluated, regardless of waiver. a `fail 'I broke this'` line or any other failure would break the control, but the most painful element is the performance hit. 

## This approach

This implementation will stop any waiver code from being executed. This will, as
a side effect, preclude all features that require skipped controls to be run. Those
features include (but are not limited to):

* Reporting number of skipped controls
* Output about results of skipped controls (sometimes we say if something skipped passed)
* `run: true` functionality no longer applies

It will, however, only run the correct controls.

This first implementation tests and focuses directly on the `waivers.yml` implementation; working on other paths is incidental and not tested in this MVP prototype.

# Current Behavior

Everything gets run behind the scenes. Even breaks in waived controllers!

```
bundle exec inspec exec test/fixtures/profiles/waivers/purely-broken-controls --waiver-file test/fixtures/profiles/waivers/purely-broken-controls/files/waivers.yml

#=>

Profile: purely-broken-controls
Version: 0.1.0
Target:  local://

  ×  01_my_broken_control: test/fixtures/profiles/waivers/purely-broken-controls/controls/purely_broken_controls.rb:1
     ×  Control Source Code Error test/fixtures/profiles/waivers/purely-broken-controls/controls/purely_broken_controls.rb:1 
     Everything is destroyed as this has been run
  ↺  02_my_working_but_waived_control: No-op
     ↺  Skipped control due to waiver condition: Sound reasoning
  ✔  03_my_working_control: true
     ✔  true is expected to eq true


Profile Summary: 1 successful control, 1 control failure, 1 control skipped
Test Summary: 1 successful, 1 failure, 1 skipped
```

# --ludicrous-speed
To test this passing, within the local directory run:

```
bundle exec inspec exec test/fixtures/profiles/waivers/purely-broken-controls --waiver-file test/fixtures/profiles/waivers/purely-broken-controls/files/waivers.yml

#=>
********************************************************************************

--ludicrous-speed is an experimental feature that stops waived controls altogether from being read into InSpec. Use with caution.

********************************************************************************
Profile: purely-broken-controls
Version: 0.1.0
Target:  local://

  ✔  03_my_working_control: true
     ✔  true is expected to eq true


Profile Summary: 1 successful control, 0 control failures, 0 controls skipped
Test Summary: 1 successful, 0 failures, 0 skipped

```

`01` control should not exist at all or be skipped. Instead it fully evaluates and fails as seen by the failure. This is bad from an output perspective but also an issue from performance. That means a `sleep 100` (or some other time based issue) would stop everything for 100 seconds. This PR fixes that.

I have included a functional test that must be passed for this to work can be run as: `bundle exec m test/functional/waivers_test.rb:118`. This PR currently passes that test.

# --retain-waiver-data

This allows a second, alternative approach, where all prior existing functionality is retained. (i.e. we still have Skip counts, etc.) However we still don't evaluate any waived control code and it does not break. Run locally:

```
bundle exec inspec exec test/fixtures/profiles/waivers/purely-broken-controls --ludicrous-speed --retain-waiver-data --waiver-file test/fixtures/profiles/waivers/purely-broken-controls/files/waivers.yml

#=>

********************************************************************************

--ludicrous-speed is an experimental feature that stops waived controls altogether from being read into InSpec. Use with caution.

********************************************************************************
Profile: purely-broken-controls
Version: 0.1.0
Target:  local://

  ↺  01_my_broken_control: No-op
     ↺  Skipped control due to waiver condition: Sound reasoning
  ↺  02_my_working_but_waived_control: No-op
     ↺  Skipped control due to waiver condition: Sound reasoning
  ✔  03_my_working_control: true
     ✔  --- true is expected to eq true


Profile Summary: 1 successful control, 0 control failures, 2 controls skipped
Test Summary: 1 successful, 0 failures, 2 skipped
```

**TODO:**

I intend on moving this logic earlier in the code read and implement on the AST/AST-autocorrect level instead of on a pure ruby-parsing level. However this currently works as seen above with noted warnings and issues. (More discussion below)

### CODE REVIEW

File: lib/inspec/profile.rb [hunk 1]
```diff
@@ -212,6 +212,36 @@ def collect_tests(include_list = @controls)
...
+        ui.red "*" * 80
+        ui.red "\n\n"
+        ui.red "--ludicrous-speed is an experimental feature that stops\
+ waived controls altogether from being read into InSpec. Use with caution."
+        ui.red "\n\n"
+        ui.red "*" * 80
+
+        ## Find the waivers file
+        collection_source_data_files = @runtime_profile.instance_variable_get(:@src).data_files
+        waiver_path = collection_source_data_files.keys.grep(/waivers.yml$/).first
```
Comment (clintoncwolfe) [inline lib/inspec/profile.rb]: So, here's a still-gross-but-a-little-safer way to do this:
```
waiver_path = Inspec::Config.cached.instance_variable_get(:@cli_opts)["waiver_file"]&.first

## Pull together waiver
if waiver_path
  waived_controls = YAML.load_file(waiver_path).keys
end
```

That's better because it reaches less, and it also does not assume that the waiver file will be named `waivers.yml`.

It still has problems:

* We could expose cli_opts and avoid the instance_variable_get
* We are only accessing the first waiver file. What if there are are multiple?
* We don't handle misparsing the waiver file.

File: lib/inspec/profile.rb [hunk 2]
```diff
@@ -212,6 +212,36 @@ def collect_tests(include_list = @controls)
 
         locked_dependencies.each(&:collect_tests)
 
+        # TODO: if --ludicrous-speed
```
Comment (clintoncwolfe) [inline lib/inspec/profile.rb]:
```
if Inspec::Config.cached["ludicrous_speed"]
 #...
end
```

Comment (Schwad) [reply lib/inspec/profile.rb]: Implemented, thanks! :)

File: lib/inspec/profile.rb (hunk 1 above)
Comment (Schwad) [reply lib/inspec/profile.rb]: Reformatted in this way, and manually referenced these issues in a comment above so we don't lose them

**Comment by Schwad** [comment]
Updated PR and description to address comments and add additional `--retain-waiver-data` flag. This flag preserves prior functionality with maintaining skip counts while retaining performance boosts of ludicrous speed. (flags must be used together)

File: lib/inspec/profile.rb [hunk 3]
```diff
@@ -213,6 +213,7 @@ def collect_tests(include_list = @controls)
         locked_dependencies.each(&:collect_tests)
 
         # Wipe out waived controls
+        require 'pry'; binding.pry
```
Comment (clintoncwolfe) [inline lib/inspec/profile.rb]: Needs to be removed - debugger

File: lib/inspec/profile.rb [hunk 4]
```diff
@@ -231,17 +230,33 @@ def collect_tests(include_list = @controls)
...
+            # waivers. This allows current waiver-data (e.g. skipis) to still
+            # be processed and rendered
+            tests.each do |key, value|
+              cleared_tests = value.split("control ").collect do |element|
+                next if element.blank?
+                splitlines = element.split("\n")
+                new_element = splitlines[0] + "\ndescribe '---' do\n" + splitlines[1..-1].join("\n") + "\nend\n"
+              end.join("control ")
+              purged_tests[key] = cleared_tests
+            end
```
Comment (clintoncwolfe) [inline lib/inspec/profile.rb]: This needs to be looked at very closely, this seems.... fishy.

Comment (Schwad) [reply lib/inspec/profile.rb]: If there's any way someone can find the lazy evaluation tactics used by RSpec in 'describe' and safely port that approach to 'control' we'd be in business.

I lost a few hours in a rabbit hole on that one.

Nothing inside a 'describe' block is evaluated when a controller is waived. (It can be tested with a 'raise' or 'sleep'). So until 'control' can be modified this would wrap all code in an unnamed 'describe' block at runtime.

I concur that the current approach is fishy- but should work in a desperate situation.

I hope I can find a few hours soon just to look at 'control' and the guts of 'describe' again.

If anyone cracks the 'control' case first let me know as I'd be very interested how it was done!

File: docs-chef-io/content/inspec/cli.md [hunk 5]
```diff
@@ -272,6 +272,8 @@ This subcommand has additional options:
     Password for enable mode on Cisco IOS devices.
 * ``--filter-empty-profiles``, ``--no-filter-empty-profiles``
     Filter empty profiles (profiles without controls) from the report.
+* ``--filter-waived-controls``
+    Do not execute waived controls in InSpec at all. Must use with --waiver-file. Ignores `run` setting of waiver file.
```
Comment (IanMadd) [inline docs-chef-io/content/inspec/cli.md:301]:
```suggestion
    Do not execute waived controls in InSpec at all. Must use with `--waiver-file`. Ignores `run` setting of waiver file.
```
